src/shadowbox/infrastructure/json_config.ts

```diff
@@ -0,0 +1,107 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as fs from 'fs';
+
+import * as file_read from './file_read';
+import * as logging from './logging';
+
+export interface JsonConfig<T> {
+  // Returns the config data as a json object.
+  data(): T;
+  // Writes the config to the backing storage.
+  write();
+}
+
+export function loadFileConfig<T>(filename: string): JsonConfig<T> {
+  const text = file_read.readFileIfExists(filename);
+  let dataJson = {} as T;
+  if (text) {
+    try {
+      dataJson = JSON.parse(text) as T;
+    } catch (error) {
+      logging.error(`Failed to parse config ${filename}: ${error}`);
+    }
+  }
+  return new FileConfig<T>(filename, dataJson);
+}
+
+// FileConfig is a JsonConfig backed by a filesystem file.
+export class FileConfig<T> implements JsonConfig<T> {
+  constructor(private filename: string, private dataJson: T) {}
+
+  data(): T {
+    return this.dataJson;
+  }
+
+  write() {
+    // Write to temporary file, then move that temporary file to the
+    // persistent location, to avoid accidentally breaking the stats file.
+    // Use *Sync calls for atomic operations, to guard against corrupting
+    // these files.
+    const tempFilename = `${this.filename}.${Date.now()}`;
+    try {
+      fs.writeFileSync(tempFilename, JSON.stringify(this.dataJson), {encoding: 'utf8'});
+      fs.renameSync(tempFilename, this.filename);
+    } catch (error) {
+      logging.error(`Error writing config ${this.filename} ${error}`);
+    }
+  }
+}
+
+// ChildConfig is a JsonConfig backed by another config.
+export class ChildConfig<T> implements JsonConfig<T> {
+  constructor(private parentConfig: JsonConfig<{}>, private dataJson: T) {}
+  data(): T {
+    return this.dataJson;
+  }
+  write() {
+    this.parentConfig.write();
+  }
+}
+
+// DelayedConfig is a JsonConfig that only writes the data in a periodic time interval.
+// Calls to write() will mark the data as "dirty" for the next inverval.
+export class DelayedConfig<T> implements JsonConfig<T> {
+  private dirty = false;
+  constructor(private config: JsonConfig<T>, writePeriodMs: number) {
+    setInterval(() => {
+      if (!this.dirty) {
+        return;
+      }
+      this.config.write();
+      this.dirty = false;
+    }, writePeriodMs);
+  }
+  data(): T {
+    return this.config.data();
+  }
+  write() {
+    this.dirty = true;
+  }
+}
+
+// InMemoryConfig is a JsonConfig backed by an internal member variable. Useful for testing.
+export class InMemoryConfig<T> implements JsonConfig<T> {
+  public written: T;
+  constructor(private dataJson: T) {
```

Comment:
I don't think there's any need for `written`:
```ts
// InMemoryConfig is a JsonConfig backed by an internal member variable. Useful for testing.
export class InMemoryConfig<T> implements JsonConfig<T> {
  constructor(private readonly config: T) {}

  data() {
    return this.config;
  }

  write() {
    JSON.parse(JSON.stringify(this.config));
  }
}
```

Reply:
OK. Not obvious! Suggest you rename `written` -> `mostRecentWrite`, or similar and add a comment.

Reply:
Renamed as per your suggestion.
Thanks for the review!